File: crates/brainwires-rag/examples/rag_search.rs
//! RAG Search — Codebase Indexing and Hybrid Semantic+Keyword Search
//!
//! Demonstrates how to initialize a `RagClient`, index a directory of
//! source files, and perform hybrid (vector + BM25 keyword) queries
//! against the indexed codebase.
//!
//! **Note:** This example requires a local codebase directory to index.
//! By default it indexes the brainwires-knowledge crate itself.
//!
//! Run:
//! ```sh
//! cargo run -p brainwires-knowledge --example rag_search --features rag
//! ```

use brainwires_rag::rag::client::RagClient;
use brainwires_rag::rag::types::{IndexRequest, QueryRequest};

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    // Initialize tracing for visibility into what the RAG system is doing
    tracing_subscriber::fmt()
        .with_max_level(tracing_subscriber::filter::LevelFilter::INFO)
        .init();

    println!("=== Brainwires RAG Search Example ===\n");

    // ── Step 1: Create the RagClient ─────────────────────────────────
    println!("--- Step 1: Initialize RagClient ---\n");

    let client = RagClient::new().await?;
    println!("RagClient initialized successfully.\n");

    // ── Step 2: Index a codebase directory ────────────────────────────
    println!("--- Step 2: Index a Codebase ---\n");

    // Index this crate's own source code as a demonstration
    let crate_dir = std::env::var("CARGO_MANIFEST_DIR").unwrap_or_else(|_| ".".to_string());
    let src_dir = std::path::Path::new(&crate_dir).join("src");
    let index_path = if src_dir.exists() {
        src_dir.display().to_string()
    } else {
        // Fallback: index current directory
        ".".to_string()
    };

    println!("Indexing directory: {}", index_path);

    let index_request = IndexRequest {
        path: index_path.clone(),
        project: Some("rag-example".to_string()),
        include_patterns: vec!["**/*.rs".to_string()],
        exclude_patterns: vec!["**/target/**".to_string(), "**/node_modules/**".to_string()],
        max_file_size: 1_048_576, // 1 MB
    };

    let index_response = client.index_codebase(index_request).await?;
    println!("Indexing complete:");
    println!("  Mode:       {:?}", index_response.mode);
    println!("  Files:      {}", index_response.files_indexed);
    println!("  Chunks:     {}", index_response.chunks_created);
    println!("  Embeddings: {}", index_response.embeddings_generated);
    println!("  Duration:   {} ms", index_response.duration_ms);
    if !index_response.errors.is_empty() {
        println!("  Errors:     {}", index_response.errors.len());
        for err in &index_response.errors {
            println!("    - {}", err);
        }
    }
    println!();

    // ── Step 3: Check statistics ─────────────────────────────────────
    println!("--- Step 3: Index Statistics ---\n");

    let stats = client.get_statistics().await?;
    println!("Total files:      {}", stats.total_files);
    println!("Total chunks:     {}", stats.total_chunks);
    println!("Total embeddings: {}", stats.total_embeddings);
    println!("Database size:    {} bytes", stats.database_size_bytes);
    println!("Language breakdown:");
    for lang in &stats.language_breakdown {
        println!(
            "  {}: {} files, {} chunks",
            lang.language, lang.file_count, lang.chunk_count
        );
    }
    println!();

    // ── Step 4: Perform hybrid semantic + keyword queries ─────────────
    println!("--- Step 4: Semantic Search Queries ---\n");

    let queries = [
        "entity relationship graph traversal",
        "embedding vector search",
        "error handling and result types",
    ];

    for query_text in &queries {
        println!("Query: \"{}\"", query_text);

        let query = QueryRequest {
            query: query_text.to_string(),
            path: Some(index_path.clone()),
            project: Some("rag-example".to_string()),
            limit: 3,
            min_score: 0.5,
            hybrid: true, // combine vector similarity with BM25 keyword scoring
        };

        let response = client.query_codebase(query).await?;
        println!(
            "  Found {} results in {} ms (threshold: {:.2}, lowered: {})",
            response.results.len(),
            response.duration_ms,
            response.threshold_used,
            response.threshold_lowered
        );

        for (i, result) in response.results.iter().enumerate() {
            let preview = result.content.lines().take(2).collect::<Vec<_>>().join(" ");
            let preview = if preview.len() > 100 {
                format!("{}...", &preview[..100])
            } else {
                preview
            };

            println!(
                "  [{}] score={:.3} (vec={:.3}, kw={}) | {}:{}-{} | {}",
                i + 1,
                result.score,
                result.vector_score,
                result
                    .keyword_score
                    .map(|s| format!("{:.3}", s))
                    .unwrap_or_else(|| "n/a".into()),
                result.file_path,
                result.start_line,
                result.end_line,
                preview
            );
        }
        println!();
    }

    // ── Step 5: Vector-only search (disable hybrid) ──────────────────
    println!("--- Step 5: Vector-Only Search ---\n");

    let query = QueryRequest {
        query: "how are thoughts stored and retrieved".to_string(),
        path: Some(index_path.clone()),
        project: Some("rag-example".to_string()),
        limit: 3,
        min_score: 0.5,
        hybrid: false, // pure vector similarity, no keyword scoring
    };

    let response = client.query_codebase(query).await?;
    println!("Vector-only results ({} found):", response.results.len());
    for (i, result) in response.results.iter().enumerate() {
        println!(
            "  [{}] score={:.3} | {} (lines {}-{})",
            i + 1,
            result.score,
            result.file_path,
            result.start_line,
            result.end_line
        );
    }

    println!("\n=== Done ===");
    Ok(())
}

File: deno/examples/knowledge/rag_search.ts
// Example: RAG Search
// Demonstrates the RagClient interface for codebase indexing and hybrid semantic+keyword search.
// Run: deno run deno/examples/cognition/rag_search.ts

import type {
  IndexRequest,
  IndexResponse,
  QueryRequest,
  QueryResponse,
  RagClient,
  SearchResult,
  StatisticsResponse,
} from "@rullama/knowledge";
import {
  DEFAULT_LIMIT,
  DEFAULT_MAX_FILE_SIZE,
  DEFAULT_MIN_SCORE,
} from "@rullama/knowledge";

// ---------------------------------------------------------------------------
// Mock RagClient — in production, supply a real vector-DB-backed implementation
// ---------------------------------------------------------------------------

class MockRagClient implements RagClient {
  private indexed = false;

  async indexCodebase(req: IndexRequest): Promise<IndexResponse> {
    console.log(
      `  [mock] Indexing ${req.path} (patterns: ${
        req.includePatterns?.join(", ") ?? "*"
      })`,
    );
    this.indexed = true;
    return {
      mode: "full",
      filesIndexed: 42,
      chunksCreated: 318,
      embeddingsGenerated: 318,
      durationMs: 1234,
      errors: [],
      filesUpdated: 0,
      filesRemoved: 0,
    };
  }

  async queryCodebase(req: QueryRequest): Promise<QueryResponse> {
    const results: SearchResult[] = [
      {
        filePath: "src/knowledge/entity.ts",
        content:
          "export interface Entity {\n  name: string;\n  entityType: EntityType;",
        score: 0.92,
        vectorScore: 0.88,
        keywordScore: req.hybrid ? 0.95 : undefined,
        startLine: 10,
        endLine: 25,
        language: "TypeScript",
        indexedAt: Date.now(),
      },
      {
        filePath: "src/rag/types.ts",
        content:
          "export interface SearchResult {\n  filePath: string;\n  score: number;",
        score: 0.85,
        vectorScore: 0.85,
        startLine: 40,
        endLine: 60,
        language: "TypeScript",
        indexedAt: Date.now(),
      },
      {
        filePath: "src/core/error.ts",
        content:
          "export class FrameworkError extends Error {\n  constructor(kind, message)",
        score: 0.78,
        vectorScore: 0.78,
        startLine: 5,
        endLine: 20,
        language: "TypeScript",
        indexedAt: Date.now(),
      },
    ];

    return {
      results: results.slice(0, req.limit ?? DEFAULT_LIMIT),
      durationMs: 23,
      thresholdUsed: req.minScore ?? DEFAULT_MIN_SCORE,
      thresholdLowered: false,
    };
  }

  async getStatistics(): Promise<StatisticsResponse> {
    return {
      totalFiles: 42,
      totalChunks: 318,
      totalEmbeddings: 318,
      databaseSizeBytes: 2_456_000,
      languageBreakdown: [
        { language: "TypeScript", fileCount: 30, chunkCount: 220 },
        { language: "Rust", fileCount: 12, chunkCount: 98 },
      ],
    };
  }

  async clearIndex() {
    this.indexed = false;
    return { success: true, message: "Index cleared" };
  }

  async advancedSearch(
    req: import("@rullama/knowledge").AdvancedSearchRequest,
  ): Promise<QueryResponse> {
    return this.queryCodebase({
      query: req.query,
      limit: req.limit,
      minScore: req.minScore,
    });
  }

  async searchGitHistory(
    req: import("@rullama/knowledge").SearchGitHistoryRequest,
  ) {
    return {
      results: [],
      commitsIndexed: 0,
      totalCachedCommits: 0,
      durationMs: 5,
    };
  }
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main() {
  console.log("=== Brainwires RAG Search Example ===\n");

  // 1. Create the RagClient
  console.log("--- Step 1: Initialize RagClient ---\n");
  const client: RagClient = new MockRagClient();
  console.log("RagClient initialized successfully.\n");

  // 2. Index a codebase directory
  console.log("--- Step 2: Index a Codebase ---\n");

  const indexPath = "./src";
  console.log(`Indexing directory: ${indexPath}`);

  const indexRequest: IndexRequest = {
    path: indexPath,
    project: "rag-example",
    includePatterns: ["**/*.ts", "**/*.rs"],
    excludePatterns: ["**/target/**", "**/node_modules/**"],
    maxFileSize: DEFAULT_MAX_FILE_SIZE,
  };

  const indexResponse = await client.indexCodebase(indexRequest);
  console.log("Indexing complete:");
  console.log(`  Mode:       ${indexResponse.mode}`);
  console.log(`  Files:      ${indexResponse.filesIndexed}`);
  console.log(`  Chunks:     ${indexResponse.chunksCreated}`);
  console.log(`  Embeddings: ${indexResponse.embeddingsGenerated}`);
  console.log(`  Duration:   ${indexResponse.durationMs} ms`);
  if (indexResponse.errors.length > 0) {
    console.log(`  Errors:     ${indexResponse.errors.length}`);
    for (const err of indexResponse.errors) {
      console.log(`    - ${err}`);
    }
  }
  console.log();

  // 3. Check statistics
  console.log("--- Step 3: Index Statistics ---\n");

  const stats = await client.getStatistics();
  console.log(`Total files:      ${stats.totalFiles}`);
  console.log(`Total chunks:     ${stats.totalChunks}`);
  console.log(`Total embeddings: ${stats.totalEmbeddings}`);
  console.log(`Database size:    ${stats.databaseSizeBytes} bytes`);
  console.log("Language breakdown:");
  for (const lang of stats.languageBreakdown) {
    console.log(
      `  ${lang.language}: ${lang.fileCount} files, ${lang.chunkCount} chunks`,
    );
  }
  console.log();

  // 4. Perform hybrid semantic + keyword queries
  console.log("--- Step 4: Semantic Search Queries ---\n");

  const queries = [
    "entity relationship graph traversal",
    "embedding vector search",
    "error handling and result types",
  ];

  for (const queryText of queries) {
    console.log(`Query: "${queryText}"`);

    const query: QueryRequest = {
      query: queryText,
      path: indexPath,
      project: "rag-example",
      limit: 3,
      minScore: 0.5,
      hybrid: true,
    };

    const response = await client.queryCodebase(query);
    console.log(
      `  Found ${response.results.length} results in ${response.durationMs} ms ` +
        `(threshold: ${
          response.thresholdUsed.toFixed(2)
        }, lowered: ${response.thresholdLowered})`,
    );

    for (let i = 0; i < response.results.length; i++) {
      const result = response.results[i];
      let preview = result.content.split("\n").slice(0, 2).join(" ");
      if (preview.length > 100) {
        preview = preview.slice(0, 100) + "...";
      }

      const kwLabel = result.keywordScore != null
        ? result.keywordScore.toFixed(3)
        : "n/a";

      console.log(
        `  [${i + 1}] score=${result.score.toFixed(3)} ` +
          `(vec=${result.vectorScore.toFixed(3)}, kw=${kwLabel}) | ` +
          `${result.filePath}:${result.startLine}-${result.endLine} | ${preview}`,
      );
    }
    console.log();
  }

  // 5. Vector-only search (disable hybrid)
  console.log("--- Step 5: Vector-Only Search ---\n");

  const vectorQuery: QueryRequest = {
    query: "how are thoughts stored and retrieved",
    path: indexPath,
    project: "rag-example",
    limit: 3,
    minScore: 0.5,
    hybrid: false,
  };

  const vectorResponse = await client.queryCodebase(vectorQuery);
  console.log(`Vector-only results (${vectorResponse.results.length} found):`);
  for (let i = 0; i < vectorResponse.results.length; i++) {
    const result = vectorResponse.results[i];
    console.log(
      `  [${i + 1}] score=${result.score.toFixed(3)} | ` +
        `${result.filePath} (lines ${result.startLine}-${result.endLine})`,
    );
  }

  console.log("\n=== Done ===");
}

await main();
